libs/contrib/Data/Nat/Factor.idr

```diff
@@ -0,0 +1,333 @@
+                rewrite remPrf in
+                Refl
+            )
+
+||| If p is a factor of n, then it is also a factor of any multiply of n.
+multNAlsoFactor : Factor n p -> (a : Nat) -> {auto aok : LTE 1 a} -> Factor (n * a) p
+multNAlsoFactor _ Z {aok} = absurd $ succNotLTEzero aok
+multNAlsoFactor (CofactorExists {n} {p} q positN prf) (S a) =
+        CofactorExists (q * S a) (lteMultRight positN a) $
+            rewrite sym prf in
+            multAssociative p q (S a)
+
+||| If p is a factor of both n and m, then it is also a factor of their sum.
+plusFactor : Factor n p -> Factor m p -> Factor (n + m) p
+plusFactor {n} {p} nFactor@(CofactorExists qn positN prfN) (CofactorExists qm positM prfM) =
+        let positP = the (LTE 1 p) $ case n of
+                Z => absurd $ succNotLTEzero positN
+                (S k) => nonZeroLeftFactor {a = p} {b = qn} prfN
+            positQNQM = the (LTE 1 (qn + qm)) $ case qn of
+                Z => absurd . succNotLTEzero .
+                    replace {P = LTE 1} (multZeroRightZero p) $
+                    replace {P = LTE 1} (sym prfN) positN
+                (S k) => LTESucc LTEZero
+        in
+        rewrite sym prfN in
+        rewrite sym prfM in
+        rewrite sym $ multDistributesOverPlusRight p qn qm in
+        multFactor p (qn + qm) {positQ = positQNQM}
+
+||| If p is a factor of a sum (n + m) and a factor of n, then it is also
+||| a factor of m. This could be expressed more naturally with minus, but
+||| it would be more difficult to prove, since minus lacks certain properties
+||| that one would expect from decent subtraction.
+minusFactor : {auto positB : LTE 1 b} -> Factor (a + b) p -> Factor a p -> Factor b p
+minusFactor {a} {b} {positB} (CofactorExists qab _ prfAB) (CofactorExists qa _ prfA) =
+        CofactorExists (minus qab qa) positB (
+            rewrite multDistributesOverMinusRight p qab qa in
+            rewrite prfA in
+            rewrite prfAB in
+            sym (
+                (b) ={ sym $ minusZeroRight b }=
+                (minus b 0) ={ sym $ plusMinusLeftCancel a b 0 }=
+                (minus (a + b) (a + 0)) ={ replace {P = \x => minus (a + b) (a + 0) = minus (a + b) x} (plusZeroRightNeutral a) Refl }=
+                (minus (a + b) a) QED
+            )
+        )
+
+||| If p is a common factor of a and b, then it is also a factor of their GCD.
+||| This actually follows directly from the definition of GCD.
+commonFactorAlsoFactorOfGCD : Factor a p -> Factor b p -> GCD a b q -> Factor q p
+commonFactorAlsoFactorOfGCD {p} pfa pfb (MkGCD _ greatest) =
+        greatest p (CommonFactorExists p pfa pfb)
+
+
+||| A decision procedure for whether of not p is a factor of n.
+decFactor : (n, d : Nat) -> {auto nok : LTE 1 n} -> {auto dok : LTE 1 d} -> DecFactor n d
+decFactor n (S d) {nok} {dok} with (Data.Fin.Extra.divMod n (S d))
+        | (Fraction n (S d) q r prf) = case r of
```

Comment:
This case statement can be pushed back into the arg list:

```idris
decFactor n (S d) {nok} {dok} with (Data.Fin.Extra.divMod n (S d))
  | (Fraction n (S d) q FZ prf) =
    let prf =
            replace {P = \x => x = n} (plusZeroRightNeutral (q + (d * q))) $
            replace {P = \x => x + 0 = n} (multCommutative q (S d)) prf
    in
    ItIsFactor $ CofactorExists q nok prf
  | (Fraction n (S d) q (FS pr) prf) =
    ItIsNotFactor $ ProperRemExists q pr nok $
      rewrite multCommutative d q in
      rewrite sym $ multRightSuccPlus q d in
      prf
```

(Indent to taste)

Reply:
Yes… I usually make a double indentation after the function pattern, because if I'm going to add a `where` clause later, I like it to be less indented than the function body. I feel it's more apparent where the body ends that way. And if I'm not going to add a `where`, at least all functions are indented according to the same rule. However, I'm not sure if it's the right approach. I also like 4-space-wide indents, while you apparently prefer 2-spaces only. I should probably have a closer look at older modules and adapt my style to what there already is…

Reply:
Mostly I stick to the indentation levels provided by the Emacs mode, and I don't think it ever provides 4-spacing. (It doesn't matter, of course.)

Reply:
Actually I think it does matter in the sense that it should be consistent within a project.